[foss/2019a] PyTorch 1.2.0 w/ Python 3.7.2

branfosj opened the pull request:
~This is a WIP. Provided to help anyone else looking at getting this going.~

boegel commented:
@branfosj Consider me interested... Certainly up for testing whenever that would be helpful.

Can you outline what's left to be done here?

branfosj commented:
I've not yet tested this on a x86_64 with GPU (our users are keeping those systems busy!). The GPU test builds I've done have been on ppc64le.

So far, I've just disabled any test on failure. I intend to go back through and see if why they are failing.

boegel commented:
@branfosj As is happens, we have an empty x86 GPU box (Cascade Lake + 4x V100) waiting for something to do. ;)

Is this PR ready to test?

boegel commented:
Test report by @boegel
**SUCCESS**
Build succeeded for 0 out of 0 (1 easyconfigs in this PR)
node3300.joltik.os - Linux centos linux 7.6.1810, Intel(R) Xeon(R) Gold 6242 CPU @ 2.80GHz, Python 2.7.5
See https://gist.github.com/f17a1a5e37422d5456fc020db7ab84c4 for a full test report.

branfosj commented:
Yes, I think this is ready to test. The `test_mul` failure, that I disable in `PyTorch-1.2.0_disable-tests.patch`, fails for me on both haswell and cascadelake foss-2019a builds of PyTorch.

boegel commented:
@branfosj How confident are you that it's OK to disable the tests that fail?
Are the PyTorch tests known to be sensitive to the processor architecture (to avoid saying "useless")?

boegel commented:
Test report by @boegel
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in this PR)
node3300.joltik.os - Linux centos linux 7.6.1810, Intel(R) Xeon(R) Gold 6242 CPU @ 2.80GHz, Python 2.7.5
See https://gist.github.com/535d7000117e45eab0e8675a6095f454 for a full test report.

branfosj commented:
I am never happy about disabling tests, but we have disabled various ones in the past - in [`PyTorch-1.0.1-foss-2018b-Python-3.6.6.eb`](https://github.com/easybuilders/easybuild-easyconfigs/blob/master/easybuild/easyconfigs/p/PyTorch/PyTorch-1.0.1-foss-2018b-Python-3.6.6.eb#L223) we have:
```
# skip test subsuites that include hanging/failing tests for now
runtest += '&& python test/run_test.py --exclude autograd cpp_extensions c10d torch --verbose'
```

The difference here is I am explicitly disabling the tests that fail one-by-one. However, in `PyTorch-1.1.0-foss-2019a-Python-3.7.2.eb` (#8636) no tests are disabled, so this feels like a step back.

boegel commented:
@branfosj Any trace upstream of others running into the same failing tests?

branfosj commented:
I've had a chance to investigate the failure I was seeing on Intel - the `test_mul` failure in `test/test_mkldnn.py`. I am seeing the failure when I build through Easybuild but not when I build direct using the Easybuild modules. The issue shows up when I add the following:
```
export CC=gcc
export CPPFLAGS="-I/rds/bear-apps/devel/2019a/branfosj-eb-pytorch/EL7/EL7-cascadelake/software/OpenBLAS/0.3.6-GCC-8.2.0-2.31.1/include -I/rds/bear-apps/devel/2019a/branfosj-eb-pytorch/EL7/EL7-cascadelake/software/ScaLAPACK/2.0.2-gompi-2019a-OpenBLAS-0.3.6/include -I/rds/bear-apps/devel/2019a/branfosj-eb-pytorch/EL7/EL7-cascadelake/software/FFTW/3.3.8-gompi-2019a/include -I/rds/bear-apps/devel/2019a/branfosj-eb-pytorch/EL7/EL7-cascadelake/software/Python/3.7.2-GCCcore-8.2.0/include -I/rds/bear-apps/devel/2019a/branfosj-eb-pytorch/EL7/EL7-cascadelake/software/MPFR/4.0.2-GCCcore-8.2.0/include -I/rds/bear-apps/devel/2019a/branfosj-eb-pytorch/EL7/EL7-cascadelake/software/GMP/6.1.2-GCCcore-8.2.0/include -I/rds/bear-apps/devel/2019a/branfosj-eb-pytorch/EL7/EL7-cascadelake/software/numactl/2.0.12-GCCcore-8.2.0/include -I/rds/bear-apps/devel/2019a/branfosj-eb-pytorch/EL7/EL7-cascadelake/software/FFmpeg/4.1.3-GCCcore-8.2.0/include -I/rds/bear-apps/devel/2019a/branfosj-eb-pytorch/EL7/EL7-cascadelake/software/gflags/2.2.2-GCCcore-8.2.0/include -I/rds/bear-apps/devel/2019a/branfosj-eb-pytorch/EL7/EL7-cascadelake/software/glog/0.4.0-GCCcore-8.2.0/include"
export CFLAGS="-O2 -ftree-vectorize -march=native -fno-math-errno"
export CXX=g++
export CXXFLAGS="-O2 -ftree-vectorize -march=native -fno-math-errno"
export LDFLAGS="-L/rds/bear-apps/devel/2019a/branfosj-eb-pytorch/EL7/EL7-cascadelake/software/GCCcore/8.2.0/lib64 -L/rds/bear-apps/devel/2019a/branfosj-eb-pytorch/EL7/EL7-cascadelake/software/GCCcore/8.2.0/lib -L/rds/bear-apps/devel/2019a/branfosj-eb-pytorch/EL7/EL7-cascadelake/software/OpenBLAS/0.3.6-GCC-8.2.0-2.31.1/lib -L/rds/bear-apps/devel/2019a/branfosj-eb-pytorch/EL7/EL7-cascadelake/software/ScaLAPACK/2.0.2-gompi-2019a-OpenBLAS-0.3.6/lib -L/rds/bear-apps/devel/2019a/branfosj-eb-pytorch/EL7/EL7-cascadelake/software/FFTW/3.3.8-gompi-2019a/lib -L/rds/bear-apps/devel/2019a/branfosj-eb-pytorch/EL7/EL7-cascadelake/software/hypothesis/4.23.4-GCCcore-8.2.0/lib -L/rds/bear-apps/devel/2019a/branfosj-eb-pytorch/EL7/EL7-cascadelake/software/psutil/5.6.3-GCCcore-8.2.0/lib -L/rds/bear-apps/devel/2019a/branfosj-eb-pytorch/EL7/EL7-cascadelake/software/Python/3.7.2-GCCcore-8.2.0/lib -L/rds/bear-apps/devel/2019a/branfosj-eb-pytorch/EL7/EL7-cascadelake/software/SciPy-bundle/2019.03-foss-2019a/lib -L/rds/bear-apps/devel/2019a/branfosj-eb-pytorch/EL7/EL7-cascadelake/software/PyYAML/5.1-GCCcore-8.2.0/lib -L/rds/bear-apps/devel/2019a/branfosj-eb-pytorch/EL7/EL7-cascadelake/software/MPFR/4.0.2-GCCcore-8.2.0/lib -L/rds/bear-apps/devel/2019a/branfosj-eb-pytorch/EL7/EL7-cascadelake/software/GMP/6.1.2-GCCcore-8.2.0/lib -L/rds/bear-apps/devel/2019a/branfosj-eb-pytorch/EL7/EL7-cascadelake/software/numactl/2.0.12-GCCcore-8.2.0/lib -L/rds/bear-apps/devel/2019a/branfosj-eb-pytorch/EL7/EL7-cascadelake/software/FFmpeg/4.1.3-GCCcore-8.2.0/lib -L/rds/bear-apps/devel/2019a/branfosj-eb-pytorch/EL7/EL7-cascadelake/software/gflags/2.2.2-GCCcore-8.2.0/lib -L/rds/bear-apps/devel/2019a/branfosj-eb-pytorch/EL7/EL7-cascadelake/software/glog/0.4.0-GCCcore-8.2.0/lib64 -L/rds/bear-apps/devel/2019a/branfosj-eb-pytorch/EL7/EL7-cascadelake/software/Pillow/6.0.0-GCCcore-8.2.0/lib"
```

branfosj commented:
Adding `CFLAGS="" CXXFLAGS=""` removes the test failure on x86_64.

That and rebuilding Python to pick up https://github.com/easybuilders/easybuild-easyconfigs/pull/8781 gets rid or most of the test failures on POWER9.

I'm still investigated the `def test_det_logdet_slogdet_batched` failure.

boegel commented:
@branfosj Maybe check if including this in the easyconfig file helps with avoiding the test failures:

```python
toolchainopts = {'vectorize': False}
```

(that will basicallu get rid of the `-ftree-vectorize`)

Which compiler options are used to build the sources when `$CXXFLAGS` & co are wiped from the environment?

branfosj commented:
```
CXX flags             :  -fvisibility-inlines-hidden -fopenmp -DUSE_FBGEMM -DUSE_QNNPACK -O2 -fPIC -Wno-narrowing -Wall -Wextra -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Wno-stringop-overflow
```

I've filed an upstream bug report on the `test_det_logdet_slogdet_batched` failure: https://github.com/pytorch/pytorch/issues/25172

I'm still looking at the other two test failures.

branfosj commented:
Adding `toolchainopts = {'vectorize': False}` did not fix the issues.

I've also noticed there is a second set of CXX flags defined:
```
   CXX flags             :  -fvisibility-inlines-hidden -fopenmp -Wnon-virtual-dtor
```
Both this set and the one in the previous message get prepended by whatever we set as `CXXFLAGS`.

I'll continue testing and see if I can narrow this down further.

branfosj commented:
Further testing show that adding:
```
toolchainopts = {'optarch': False}
```
fix the failing test_mul in the mkldnn tests.

boegel commented:
@branfosj OK, that's a good lead... That implies disabling `-march=native` though, which likely implies significant performance impact.

On which processor architecture are you seeing those failures? Let me guess... Skylake?

branfosj commented:
It fails on both Cascadelake and Haswell.

I've done a bunch of testing and it seems something is going wrong with MKLDNN multiplication when using `-march=native`. I've upstreamed the bug report to PyTorch: https://github.com/pytorch/pytorch/issues/25242

branfosj commented:
Update:

- pytorch/pytorch#25242 is being worked on and a fix should arrive to stop us needing to do the `-march=native`
- pytorch/pytorch#25172 is still being investigated, but it is looking like the failure is being caused by numerical instability in a matrix computation between the reference solution and the computed solution.
- the other two tests I disable when building on POWER9 (the `relu6` and `hardtanh`) fail during the testing when run as part of the easybuild process. However, if I disable the tests, build PyTorch, and then run the tests on the installed PyTorch they succeed.

verdurin commented:
We would be interested in this one and may be able to provide more test builds, if that's helpful.

branfosj commented:
Test report by @branfosj
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in this PR)
bear-pg0211u03a.bear.cluster - Linux centos linux 7.6.1810, Intel(R) Xeon(R) Gold 6248 CPU @ 2.50GHz, Python 2.7.5
See https://gist.github.com/180fde53da35e23b01850ecf45c3e27b for a full test report.

branfosj commented:
I've added a patch to fix the issue that necessitated turning off `march=native` - this is based on the upstream one and modified both easyconfigs to remove the workaround. Test report for `PyTorch-1.2.0-foss-2019a-Python-3.7.2.eb` done and one should arrive for the CUDA version later today.

branfosj commented:
Test report by @branfosj
**SUCCESS**
Build succeeded for 3 out of 3 (3 easyconfigs in this PR)
bear-pg0212u15b.bear.cluster - Linux centos linux 7.6.1810, Intel(R) Xeon(R) CPU E5-2640 v4 @ 2.40GHz, Python 2.7.5
See https://gist.github.com/d58f7190ba2a8815211aa9d0a12df99c for a full test report.

verdurin commented:
@branfosj One thing confusing me here is that I'm running the following build:

'eb --from-pr=8753 --debug --robot PyTorch-1.2.0-foss-2019a-Python-3.7.2.eb'

However, it still brings in `CUDA` and `cuDNN`.

branfosj commented:
@verdurin Not sure. However, I've split out the fosscuda version (and magma) into #8891 - so there is now nothing here that should cause that problem.

branfosj commented on easybuild/easyconfigs/p/PyTorch/PyTorch-1.2.0_disable-tests.patch:
```diff
@@ -0,0 +1,18 @@
+--- test/test_nn.py.orig	2019-02-10 14:13:51.665671250 +0000
```
done

boegel commented:
Test report by @boegel
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in this PR)
generoso - Linux centos linux 7.6.1810, Intel(R) Xeon(R) CPU E5-2660 v4 @ 2.00GHz, Python 3.6.8
See https://gist.github.com/cdc480a30af4d5d066b6643ac17e63d3 for a full test report.

boegel commented:
Test report by @boegel
**FAILED**
Build succeeded for 0 out of 1 (1 easyconfigs in this PR)
node2065.delcatty.os - Linux centos linux 7.6.1810, Intel(R) Xeon(R) CPU E5-2670 0 @ 2.60GHz, Python 2.7.5
See https://gist.github.com/459234894d8d7428982f22bfd6f231c1 for a full test report.

boegel commented:
Test report by @boegel
**FAILED**
Build succeeded for 0 out of 1 (1 easyconfigs in this PR)
node3110.skitty.os - Linux centos linux 7.6.1810, Intel(R) Xeon(R) Gold 6140 CPU @ 2.30GHz, Python 2.7.5
See https://gist.github.com/6c146a04c3b42a9ab3a868c46d79fb05 for a full test report.

boegel commented:
Test report by @boegel
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in this PR)
node2637.swalot.os - Linux centos linux 7.6.1810, Intel(R) Xeon(R) CPU E5-2660 v3 @ 2.60GHz, Python 2.7.5
See https://gist.github.com/809d061f1a2e2582cd07e12f882e62e0 for a full test report.

boegel commented:
@branfosj The failing test reports are for partial nodes, problems could be caused because of lack of sufficient memory or cgroups, will submit another test report to try and rule that out...

boegel commented:
Test report by @boegel
**FAILED**
Build succeeded for 0 out of 1 (1 easyconfigs in this PR)
node2065.delcatty.os - Linux centos linux 7.6.1810, Intel(R) Xeon(R) CPU E5-2670 0 @ 2.60GHz, Python 2.7.5
See https://gist.github.com/871dc251da6e5ae1be2aed29386ba815 for a full test report.

boegel commented:
@branfosj Test report at https://github.com/easybuilders/easybuild-easyconfigs/pull/8753#issuecomment-531489248 was collected on a full Intel Sandy Bridge nodes (16 cores, 64GB of RAM), yet tests still fail (well, one did).

Any ideas?

branfosj commented:
@boegel I've set off a build on a Intel Sandy Bridge (16 cores, 64GB of RAM) for comparison, but it'll be a while as I've got to build all the dependencies. A quick look at the test https://github.com/pytorch/pytorch/blob/v1.2.0/test/test_mkldnn.py#L336 doesn't suggest memory should be an issue.

branfosj commented:
Test report by @branfosj
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in this PR)
bear-pg0211u12a.bear.cluster - Linux centos linux 7.6.1810, Intel(R) Xeon(R) Gold 6248 CPU @ 2.50GHz, Python 2.7.5
See https://gist.github.com/157a37e9691889e285903780e1ca0f09 for a full test report.

branfosj commented:
Test report by @branfosj
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in this PR)
bber0501u03a.bb2.cluster - Linux centos linux 7.6.1810, Intel(R) Xeon(R) CPU E5-2690 v3 @ 2.60GHz, Python 2.7.5
See https://gist.github.com/39a360ed49121d03bb219fa75f1519bb for a full test report.

branfosj commented:
Test report by @branfosj
**FAILED**
Build succeeded for 0 out of 1 (1 easyconfigs in this PR)
bb2c1a09.bb2.cluster - Linux centos linux 7.6.1810, Intel(R) Xeon(R) CPU E5-2660 0 @ 2.20GHz, Python 2.7.5
See https://gist.github.com/547e39b0de6e8b96595faa221405fdc8 for a full test report.

boegel commented:
Test report by @boegel
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in this PR)
node3153.skitty.os - Linux centos linux 7.6.1810, Intel(R) Xeon(R) Gold 6140 CPU @ 2.30GHz, Python 2.7.5
See https://gist.github.com/bee121211809c75654e68d29b0037c7f for a full test report.

boegel commented:
@branfosj Same problem as what I got in your last test report. My last test report was on Intel SKylake (full node), no problem there though.

This is starting to smell as a quirky test which fails randomly?

branfosj commented:
Both failures are on Sandy Bridge - I am wondering if there is some optimisation that does not have a suitable fallback for older processors. I'll investigate further.

branfosj commented:
I've filed a bug report upstream: https://github.com/pytorch/pytorch/issues/26287 - it seems to only happen with `-march=native` and on Sandybridge.

boegel commented:
@branfosj Do you think it's acceptable to comment out that test and get this merged?

branfosj commented:
I think so. To hit the problem you have to be using the `torch.nn.Softmax` function, pass to using the mkldnn code, and be running on an Intel SandyBridge. Even then it fails in a clear way - i.e. it does not produce a wrong answer, it produces an error message.

branfosj commented:
Test report by @branfosj
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in this PR)
bear-pg0211u03a.bear.cluster - Linux centos linux 7.6.1810, Intel(R) Xeon(R) Gold 6248 CPU @ 2.50GHz, Python 2.7.5
See https://gist.github.com/35497e0328369859c5fcb7f155b7f7e2 for a full test report.

branfosj commented:
Test report by @branfosj
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in this PR)
bber0501u03b.bb2.cluster - Linux centos linux 7.6.1810, Intel(R) Xeon(R) CPU E5-2690 v3 @ 2.60GHz, Python 2.7.5
See https://gist.github.com/7a5368ed95ba1180f4f62f77e9c03858 for a full test report.

branfosj commented:
Test report by @branfosj
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in this PR)
bb2c1a09.bb2.cluster - Linux centos linux 7.6.1810, Intel(R) Xeon(R) CPU E5-2660 0 @ 2.20GHz, Python 2.7.5
See https://gist.github.com/6c02749749e39916c3b20f57430ca964 for a full test report.

boegel commented:
Test report by @boegel
**SUCCESS**
Build succeeded for 11 out of 11 (1 easyconfigs in this PR)
node3120.skitty.os - Linux centos linux 7.6.1810, Intel(R) Xeon(R) Gold 6140 CPU @ 2.30GHz, Python 2.7.5
See https://gist.github.com/3c772ac8e6d72acc8e6d67ac31a007fb for a full test report.

boegel commented:
Test report by @boegel
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in this PR)
node2400.golett.os - Linux centos linux 7.6.1810, Intel(R) Xeon(R) CPU E5-2680 v3 @ 2.50GHz, Python 2.7.5
See https://gist.github.com/14e94cc615e178cd327c1c03546daa97 for a full test report.

boegel commented:
Going in, thanks @branfosj!

boegel commented:
Test report by @boegel
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in this PR)
node2105.delcatty.os - Linux centos linux 7.6.1810, Intel(R) Xeon(R) CPU E5-2670 0 @ 2.60GHz, Python 2.7.5
See https://gist.github.com/9ef982864585c3c3a5f5492a25c3e80a for a full test report.
